src/pool-bin/libraries/BinHooks.sol

```diff
@@ -0,0 +1,158 @@
+        callerDelta = delta;
+
+        if (key.parameters.shouldCall(HOOKS_AFTER_BURN_OFFSET, hooks)) {
+            bytes4 selector;
+            (selector, hookDelta) = hooks.afterBurn(msg.sender, key, params, delta, hookData);
+
+            if (selector != IBinHooks.afterBurn.selector) {
+                revert Hooks.InvalidHookResponse();
+            }
+
+            if (
+                key.parameters.hasOffsetEnabled(HOOKS_AFTER_BURN_RETURNS_DELTA_OFFSET)
+                    && hookDelta != BalanceDeltaLibrary.ZERO_DELTA
+            ) {
+                callerDelta = callerDelta - hookDelta;
+            }
+        }
+    }
+
+    function beforeSwap(PoolKey memory key, bool swapForY, uint128 amountIn, bytes calldata hookData)
+        internal
+        returns (uint128 amountToSwap, int128 hookDeltaSpecified)
+    {
+        IBinHooks hooks = IBinHooks(address(key.hooks));
+        amountToSwap = amountIn;
+
+        /// @notice If the hook is not registered, return the original amount to swap
+        if (!key.parameters.shouldCall(HOOKS_BEFORE_SWAP_OFFSET, hooks)) {
+            return (amountToSwap, hookDeltaSpecified);
+        }
+
+        bytes4 selector;
+        (selector, hookDeltaSpecified) = hooks.beforeSwap(msg.sender, key, swapForY, amountIn, hookData);
+        if (selector != IBinHooks.beforeSwap.selector) {
+            revert Hooks.InvalidHookResponse();
+        }
+
+        // Update the swap amount according to the hook's return
+        if (key.parameters.hasOffsetEnabled(HOOKS_BEFORE_SWAP_RETURNS_DELTA_OFFSET) && hookDeltaSpecified != 0) {
+            /// @dev default overflow check make sure the swap amount is always valid
+            if (hookDeltaSpecified > 0) {
+                amountToSwap += uint128(hookDeltaSpecified);
+            } else {
+                amountToSwap -= uint128(-hookDeltaSpecified);
+            }
+        }
+    }
+
+    function afterSwap(
+        PoolKey memory key,
+        bool swapForY,
+        uint128 amountIn,
+        BalanceDelta delta,
+        bytes calldata hookData,
+        int128 hookDeltaSpecified
+    ) internal returns (BalanceDelta swapperDelta, BalanceDelta hookDelta) {
+        IBinHooks hooks = IBinHooks(address(key.hooks));
+        swapperDelta = delta;
+
+        int128 hookDeltaUnspecified;
+        if (key.parameters.shouldCall(HOOKS_AFTER_SWAP_OFFSET, hooks)) {
+            bytes4 selector;
+            (selector, hookDeltaUnspecified) = hooks.afterSwap(msg.sender, key, swapForY, amountIn, delta, hookData);
+            if (selector != IBinHooks.afterSwap.selector) {
+                revert Hooks.InvalidHookResponse();
+            }
+
+            if (!key.parameters.hasOffsetEnabled(HOOKS_AFTER_SWAP_RETURNS_DELTA_OFFSET)) {
+                hookDeltaUnspecified = 0;
+            }
+        }
+
+        if (hookDeltaUnspecified != 0 || hookDeltaSpecified != 0) {
+            hookDelta = swapForY
```

Comment:
curious why here is only checking for `swapForY` while CLPool is checking amountSpecified as well `(params.amountSpecified > 0 == params.zeroForOne)`

Reply:
exactly 😄